File: Src/Compilers/VisualBasic/Source/Analysis/FlowAnalysis/DataFlowAnalysis.vb
' Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports System
Imports System.Collections.Generic
Imports System.Threading
Imports Microsoft.CodeAnalysis.Text
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic

    ''' <summary>
    ''' This class implements the region data flow analysis operations.  Region data flow analysis provides information
    ''' how data flows into and out of a region.  The analysis is done lazily. When created, it performs no analysis, but
    ''' simply caches the arguments. Then, the first time one of the analysis results is used it computes that one
    ''' result and caches it. Each result is computed using a custom algorithm.
    ''' </summary>
    Friend Class VisualBasicDataFlowAnalysis
        Inherits DataFlowAnalysis

        Private ReadOnly _context As RegionAnalysisContext

        Private _variablesDeclared As IEnumerable(Of Symbol)
        Private _unassignedVariables As HashSet(Of Symbol)
        Private _dataFlowsIn As HashSet(Of Symbol)
        Private _dataFlowsOut As HashSet(Of Symbol)
        Private _alwaysAssigned As IEnumerable(Of Symbol)
        Private _readInside As IEnumerable(Of Symbol)
        Private _writtenInside As IEnumerable(Of Symbol)
        Private _readOutside As IEnumerable(Of Symbol)
        Private _writtenOutside As IEnumerable(Of Symbol)
        Private _captured As IEnumerable(Of Symbol)
        Private _succeeded As Boolean?
        Private _invalidRegionDetected As Boolean

        Friend Sub New(_context As RegionAnalysisContext)
            Me._context = _context
        End Sub

        ''' <summary>
        ''' A collection of the local variables that are declared within the region. Note that the region must be 
        ''' bounded by a method's body or a field's initializer, so parameter symbols are never included in the result.
        ''' </summary>
        Public Overrides ReadOnly Property VariablesDeclared As IEnumerable(Of ISymbol)
            Get
                If _variablesDeclared Is Nothing Then
                    Dim result = If(Me._context.Failed, Enumerable.Empty(Of Symbol)(),
                                    VariablesDeclaredWalker.Analyze(_context.AnalysisInfo, _context.RegionInfo))
                    Interlocked.CompareExchange(_variablesDeclared, result, Nothing)
                End If
                Return _variablesDeclared
            End Get
        End Property

        Private ReadOnly Property UnassignedVariables As HashSet(Of Symbol)
            Get
                If _unassignedVariables Is Nothing Then
                    Dim result = If(Me._context.Failed, New HashSet(Of Symbol)(),
                                    UnassignedVariablesWalker.Analyze(_context.AnalysisInfo))
                    Interlocked.CompareExchange(_unassignedVariables, result, Nothing)
                End If
                Return _unassignedVariables
            End Get
        End Property

        ''' <summary>
        ''' A collection of the local variables for which a value assigned outside the region may be used inside the region.
        ''' </summary>
        Public Overrides ReadOnly Property DataFlowsIn As IEnumerable(Of ISymbol)
            Get
                If _dataFlowsIn Is Nothing Then
                    Me._succeeded = Not Me._context.Failed
                    Dim result = If(Me._context.Failed, New HashSet(Of Symbol)(),
                                    DataFlowsInWalker.Analyze(_context.AnalysisInfo, _context.RegionInfo, UnassignedVariables, _succeeded, _invalidRegionDetected))
                    Interlocked.CompareExchange(_dataFlowsIn, result, Nothing)
                End If
                Return _dataFlowsIn
            End Get
        End Property

        ''' <summary>
        ''' A collection of the local variables for which a value assigned inside the region may be used outside the region.
        ''' </summary>
        Public Overrides ReadOnly Property DataFlowsOut As IEnumerable(Of ISymbol)
            Get
                Dim discarded = DataFlowsIn
                If _dataFlowsOut Is Nothing Then
                    Dim result = If(Me._context.Failed, New HashSet(Of Symbol)(),
                                    DataFlowsOutWalker.Analyze(_context.AnalysisInfo, _context.RegionInfo, UnassignedVariables, _dataFlowsIn))
                    Interlocked.CompareExchange(_dataFlowsOut, result, Nothing)
                End If
                Return _dataFlowsOut
            End Get
        End Property

        ''' <summary>
        ''' A collection of the local variables for which a value is always assigned inside the region.
        ''' </summary>
        Public Overrides ReadOnly Property AlwaysAssigned As IEnumerable(Of ISymbol)
            Get
                If _alwaysAssigned Is Nothing Then
                    Dim result = If(Me._context.Failed, Enumerable.Empty(Of Symbol)(),
                                    AlwaysAssignedWalker.Analyze(_context.AnalysisInfo, _context.RegionInfo))
                    Interlocked.CompareExchange(_alwaysAssigned, result, Nothing)
                End If
                Return _alwaysAssigned
            End Get
        End Property

        ''' <summary>
        ''' A collection of the local variables that are read inside the region.
        ''' </summary>
        Public Overrides ReadOnly Property ReadInside As IEnumerable(Of ISymbol)
            Get
                If _readInside Is Nothing Then
                    AnalyzeReadWrite()
                End If
                Return _readInside
            End Get
        End Property

        ''' <summary>
        ''' A collection of local variables that are written inside the region.
        ''' </summary>
        Public Overrides ReadOnly Property WrittenInside As IEnumerable(Of ISymbol)
            Get
                If _writtenInside Is Nothing Then
                    AnalyzeReadWrite()
                End If
                Return _writtenInside
            End Get
        End Property

        ''' <summary>
        ''' A collection of the local variables that are read outside the region.
        ''' </summary>
        Public Overrides ReadOnly Property ReadOutside As IEnumerable(Of ISymbol)
            Get
                If _readOutside Is Nothing Then
                    AnalyzeReadWrite()
                End If
                Return _readOutside
            End Get
        End Property

        ''' <summary>
        ''' A collection of local variables that are written inside the region.
        ''' </summary>
        Public Overrides ReadOnly Property WrittenOutside As IEnumerable(Of ISymbol)
            Get
                If _writtenOutside Is Nothing Then
                    AnalyzeReadWrite()
                End If
                Return _writtenOutside
            End Get
        End Property

        Private Sub AnalyzeReadWrite()
            Dim readInside As IEnumerable(Of Symbol) = Nothing
            Dim writtenInside As IEnumerable(Of Symbol) = Nothing
            Dim readOutside As IEnumerable(Of Symbol) = Nothing
            Dim writtenOutside As IEnumerable(Of Symbol) = Nothing
            Dim captured As IEnumerable(Of Symbol) = Nothing

            If Not Me.Succeeded Then
                readInside = Enumerable.Empty(Of Symbol)()
                writtenInside = readInside
                readOutside = readInside
                writtenOutside = readInside
                captured = readInside

            Else
                ReadWriteWalker.Analyze(
                    _context.AnalysisInfo, _context.RegionInfo,
                    readInside:=readInside,
                    writtenInside:=writtenInside,
                    readOutside:=readOutside,
                    writtenOutside:=writtenOutside,
                    captured:=captured)
            End If

            Interlocked.CompareExchange(Me._readInside, readInside, Nothing)
            Interlocked.CompareExchange(Me._writtenInside, writtenInside, Nothing)
            Interlocked.CompareExchange(Me._readOutside, readOutside, Nothing)
            Interlocked.CompareExchange(Me._writtenOutside, writtenOutside, Nothing)
            Interlocked.CompareExchange(Me._captured, captured, Nothing)
        End Sub

        ''' <summary>
        ''' A collection of the local variables that have been referenced in anonymous functions
        ''' and therefore must be moved to a field of a frame class.
        ''' </summary>
        Public Overrides ReadOnly Property Captured As IEnumerable(Of ISymbol)
            Get
                If Me._captured Is Nothing Then
                    AnalyzeReadWrite()
                End If

                Return Me._captured
            End Get
        End Property

        Friend ReadOnly Property InvalidRegionDetectedInternal As Boolean
            Get
                Return If(Me.Succeeded, False, Me._invalidRegionDetected)
            End Get
        End Property

        Public NotOverridable Overrides ReadOnly Property Succeeded As Boolean
            Get
                If Me._succeeded Is Nothing Then
                    Dim discarded = DataFlowsIn
                End If

                Return Me._succeeded.Value
            End Get
        End Property

        Public Overrides ReadOnly Property UnsafeAddressTaken As IEnumerable(Of ISymbol)
            Get
                Return Enumerable.Empty(Of ISymbol)
            End Get
        End Property
    End Class

End Namespace

File: Src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowAnalysis.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Threading;

namespace Microsoft.CodeAnalysis.CSharp
{
    /// <summary>
    /// This class implements the region data flow analysis operations.  Region data flow analysis
    /// provides information how data flows into and out of a region.  The analysis is done lazily.
    /// When created, it performs no analysis, but simply caches the arguments. Then, the first time
    /// one of the analysis results is used it computes that one result and caches it. Each result
    /// is computed using a custom algorithm.
    /// </summary>
    internal class CSharpDataFlowAnalysis : DataFlowAnalysis
    {
        private readonly RegionAnalysisContext context;

        private ImmutableArray<ISymbol> variablesDeclared;
        private HashSet<Symbol> unassignedVariables;
        private ImmutableArray<ISymbol> dataFlowsIn;
        private ImmutableArray<ISymbol> dataFlowsOut;
        private ImmutableArray<ISymbol> alwaysAssigned;
        private ImmutableArray<ISymbol> readInside;
        private ImmutableArray<ISymbol> writtenInside;
        private ImmutableArray<ISymbol> readOutside;
        private ImmutableArray<ISymbol> writtenOutside;
        private ImmutableArray<ISymbol> captured;
        private ImmutableArray<ISymbol> unsafeAddressTaken;
        private HashSet<PrefixUnaryExpressionSyntax> unassignedVariableAddressOfSyntaxes;
        private bool? succeeded = null;

        internal CSharpDataFlowAnalysis(RegionAnalysisContext context)
        {
            this.context = context;
        }

        /// <summary>
        /// A collection of the local variables that are declared within the region. Note that the region must be
        /// bounded by a method's body or a field's initializer, so method parameter symbols are never included
        /// in the result, but lambda parameters might appear in the result.
        /// </summary>
        public override ImmutableArray<ISymbol> VariablesDeclared
        {
            // Variables declared in the region is computed by a simple scan.
            // ISSUE: are these only variables declared at the top level in the region,
            // or are we to include variables declared in deeper scopes within the region?
            get
            {
                if (variablesDeclared.IsDefault)
                {
                    var result = Succeeded
                        ? ((IEnumerable<ISymbol>)VariablesDeclaredWalker.Analyze(context.Compilation, context.Member, context.BoundNode, context.FirstInRegion, context.LastInRegion)).ToImmutableArray()
                        : ImmutableArray<ISymbol>.Empty;
                    ImmutableInterlocked.InterlockedInitialize(ref variablesDeclared, result);
                }

                return variablesDeclared;
            }
        }

        private HashSet<Symbol> UnassignedVariables
        {
            get
            {
                if (unassignedVariables == null)
                {
                    var result = Succeeded
                        ? UnassignedVariablesWalker.Analyze(context.Compilation, context.Member, context.BoundNode)
                        : new HashSet<Symbol>();
                    Interlocked.CompareExchange(ref unassignedVariables, result, null);
                }

                return unassignedVariables;
            }
        }

        /// <summary>
        /// A collection of the local variables for which a value assigned outside the region may be used inside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> DataFlowsIn
        {
            get
            {
                if (dataFlowsIn == null)
                {
                    succeeded = !context.Failed;
                    var result = context.Failed ? ImmutableArray<ISymbol>.Empty :
                        ((IEnumerable<ISymbol>)DataFlowsInWalker.Analyze(context.Compilation, context.Member, context.BoundNode, context.FirstInRegion, context.LastInRegion, UnassignedVariables, UnassignedVariableAddressOfSyntaxes, out succeeded)).ToImmutableArray();
                    ImmutableInterlocked.InterlockedInitialize(ref dataFlowsIn, result);
                }

                return dataFlowsIn;
            }
        }

        /// <summary>
        /// A collection of the local variables for which a value assigned inside the region may be used outside the region.
        /// Note that every reachable assignment to a ref or out variable will be included in the results.
        /// </summary>
        public override ImmutableArray<ISymbol> DataFlowsOut
        {
            get
            {
                var discarded = DataFlowsIn; // force DataFlowsIn to be computed
                if (dataFlowsOut == null)
                {
                    var result = Succeeded
                        ? ((IEnumerable<ISymbol>)DataFlowsOutWalker.Analyze(context.Compilation, context.Member, context.BoundNode, context.FirstInRegion, context.LastInRegion, UnassignedVariables, dataFlowsIn)).ToImmutableArray()
                        : ImmutableArray<ISymbol>.Empty;
                    ImmutableInterlocked.InterlockedInitialize(ref dataFlowsOut, result);
                }

                return dataFlowsOut;
            }
        }

        /// <summary>
        /// A collection of the local variables for which a value is always assigned inside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> AlwaysAssigned
        {
            get
            {
                if (alwaysAssigned == null)
                {
                    var result = Succeeded
                        ? ((IEnumerable<ISymbol>)AlwaysAssignedWalker.Analyze(context.Compilation, context.Member, context.BoundNode, context.FirstInRegion, context.LastInRegion)).ToImmutableArray()
                        : ImmutableArray<ISymbol>.Empty;
                    ImmutableInterlocked.InterlockedInitialize(ref alwaysAssigned, result);
                }

                return alwaysAssigned;
            }
        }

        /// <summary>
        /// A collection of the local variables that are read inside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> ReadInside
        {
            get
            {
                if (readInside == null)
                {
                    AnalyzeReadWrite();
                }

                return readInside;
            }
        }

        /// <summary>
        /// A collection of local variables that are written inside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> WrittenInside
        {
            get
            {
                if (writtenInside == null)
                {
                    AnalyzeReadWrite();
                }

                return writtenInside;
            }
        }

        /// <summary>
        /// A collection of the local variables that are read outside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> ReadOutside
        {
            get
            {
                if (readOutside == null)
                {
                    AnalyzeReadWrite();
                }

                return readOutside;
            }
        }

        /// <summary>
        /// A collection of local variables that are written outside the region.
        /// </summary>
        public override ImmutableArray<ISymbol> WrittenOutside
        {
            get
            {
                if (writtenOutside == null)
                {
                    AnalyzeReadWrite();
                }

                return writtenOutside;
            }
        }

        private void AnalyzeReadWrite()
        {
            ImmutableArray<ISymbol> readInside, writtenInside, readOutside, writtenOutside, captured, unsafeAddressTaken;
            if (Succeeded)
            {
                ReadWriteWalker.Analyze(context.Compilation, context.Member, context.BoundNode, context.FirstInRegion, context.LastInRegion, UnassignedVariableAddressOfSyntaxes,
                    readInside: out readInside, writtenInside: out writtenInside,
                    readOutside: out readOutside, writtenOutside: out writtenOutside,
                    captured: out captured, unsafeAddressTaken: out unsafeAddressTaken);
            }
            else
            {
                readInside = writtenInside = readOutside = writtenOutside = captured = unsafeAddressTaken = ImmutableArray<ISymbol>.Empty;
            }

            ImmutableInterlocked.InterlockedInitialize(ref this.readInside, readInside);
            ImmutableInterlocked.InterlockedInitialize(ref this.writtenInside, writtenInside);
            ImmutableInterlocked.InterlockedInitialize(ref this.readOutside, readOutside);
            ImmutableInterlocked.InterlockedInitialize(ref this.writtenOutside, writtenOutside);
            ImmutableInterlocked.InterlockedInitialize(ref this.captured, captured);
            ImmutableInterlocked.InterlockedInitialize(ref this.unsafeAddressTaken, unsafeAddressTaken);
        }

        /// <summary>
        /// A collection of the non-constant local variables and parameters that have been referenced in anonymous functions
        /// and therefore must be moved to a field of a frame class.
        /// </summary>
        public override ImmutableArray<ISymbol> Captured
        {
            get
            {
                if (this.captured == null)
                {
                    AnalyzeReadWrite();
                }

                return this.captured;
            }
        }

        /// <summary>
        /// A collection of the non-constant local variables and parameters that have had their address (or the address of one
        /// of their fields) taken using the '&amp;' operator.
        /// </summary>
        /// <remarks>
        /// If there are any of these in the region, then a method should not be extracted.
        /// </remarks>
        public override ImmutableArray<ISymbol> UnsafeAddressTaken
        {
            get
            {
                if (this.unsafeAddressTaken == null)
                {
                    AnalyzeReadWrite();
                }

                return this.unsafeAddressTaken;
            }
        }

        private HashSet<PrefixUnaryExpressionSyntax> UnassignedVariableAddressOfSyntaxes
        {
            get
            {
                if (unassignedVariableAddressOfSyntaxes == null)
                {
                    var result = Succeeded
                        ? UnassignedAddressTakenVariablesWalker.Analyze(context.Compilation, context.Member, context.BoundNode)
                        : new HashSet<PrefixUnaryExpressionSyntax>();
                    Interlocked.CompareExchange(ref unassignedVariableAddressOfSyntaxes, result, null);
                }

                return unassignedVariableAddressOfSyntaxes;
            }
        }

        /// <summary>
        /// Returns true iff analysis was successful.  Analysis can fail if the region does not properly span a single expression,
        /// a single statement, or a contiguous series of statements within the enclosing block.
        /// </summary>
        public sealed override bool Succeeded
        {
            get
            {
                if (succeeded == null)
                {
                    var discarded = DataFlowsIn;
                }

                return succeeded.Value;
            }
        }
    }
}